### PR TITLE
Optimize performance of maps

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR optimizes the compiler's use of objects as maps as follows:

* Require the host to be ECMAScript 5 compliant and create maps using `Object.create(null)`. This ensures maps have no prototype (and thus no inherited members).
* Apply the `delete` operator to every newly created map object. This causes V8 to put the object in "dictionary mode" and disables creation of hidden classes which are very expensive for objects that are constantly changing shape.
* Skip `hasOwnProperty` calls on property lookups. They're no longer necessary when objects are known to have no prototype.

The PR introduces a new `MapLike<T>` type that represents a regular object used as a lookup (e.g. an object literal). A `MapLike<T>` requires use of `hasOwnProperty` to determine whether a particular property was inherited from the prototype.

The `Map<T>` type now has a brand and `Map<T>` instances can only be created by calling `createMap<T>()` (which in turn calls `Object.create(null)` and puts the object into dictionary mode).

Performance impact on the Monaco real world code base is as follows:

* 15% overall speed increase (6.7s vs. 7.9s).
* 19% less memory consumed (211Mb vs. 261Mb).

Definitely a worthwhile optimization, provided we're ok with dropping support for ECMAScript 3 hosts.

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
:+1:, logged #10278

**Comment by rbuckton** [comment]
@ahejlsberg #10298 is an alternative to this and is a port to master of what I was working on for the transforms branch. It has the same improvements in speed and memory usage as this PR but also keeps support for ES3.

**Comment by rbuckton** [comment]
👍 following the discussion in today's design meeting. I've closed #10298 in favor of this. I'll follow up with a separate PR for the isolated cases that should also be switched from `MapLike` to `Map`.

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -1084,7 +1084,7 @@ namespace ts {
 
         function internIdentifier(text: string): string {
             text = escapeIdentifier(text);
-            return hasProperty(identifiers, text) ? identifiers[text] : (identifiers[text] = text);
+            return identifiers[text] || (identifiers[text] = text);
```
Comment (RyanCavanaugh) [inline src/compiler/parser.ts]: Could do `=== undefined` to avoid re-interning `""`

**Comment by WanderWang** [comment]
from this PR , my project got lots of improvement on compile time and memory used .

before :



after:

158b6ebb-d3ba-4d47-9737-9d0bc97fbf3f

**Pull request merged**